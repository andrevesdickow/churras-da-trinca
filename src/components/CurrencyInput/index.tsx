'use client';

import ReactCurrencyInput from 'react-currency-input';
import { useController } from 'react-hook-form';
import { omit } from 'lodash';
import { helperTextClassName } from '../TextField';

type CurrencyInputProps = {
  name: string;
  control: any;
  label: React.ReactNode | string;
  helperText?: string;
  value?: number;
}

export const CurrencyInput = ({
  name,
  label,
  helperText,
  control,
  ...rest
}: CurrencyInputProps) => {
  const { field, fieldState: { invalid, error } } = useController({
    name,
    control,
    defaultValue: rest.value || 0
  });

  const handleChangeInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    const val = event.target.value || '0';

    const [first, second] = val.split(',');
    const result = parseFloat(`${first.replace('.', '')}.${second}`);

    field.onChange(result);
  };

  return (
    <div className="relative z-0 mb-4">
      <ReactCurrencyInput
        {...omit(field, 'ref')}
        id={name}
        className="block py-2.5 px-0 w-full text-sm text-slate-900 bg-transparent border-0 border-b-2 border-slate-500 appearance-none dark:border-slate-300 dark:focus:border-amber-400 dark:text-slate-100 focus:outline-none focus:ring-0 focus:border-amber-400 peer"
        {...rest}
        type="tel"
        onChangeEvent={handleChangeInput}
        allowEmpty={false}
        decimalSeparator=","
        thousandSeparator="."
        placeholder=" "
      />
      <label
        htmlFor={name}
        className="absolute text-sm text-slate-500 dark:text-slate-300 duration-300 transform -translate-y-6 scale-75 top-3 -z-10 origin-[0] peer-focus:left-0 peer-focus:text-amber-400 peer-focus:dark:text-amber-400 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-6"
      >
        {label}
      </label>

      {(helperText || invalid) && (
        <div className={helperTextClassName({ isInvalid: invalid })} role={invalid ? 'alert' : 'note'}>
          {invalid ? error?.message : helperText}
        </div>
      )}
    </div>
  );
};
