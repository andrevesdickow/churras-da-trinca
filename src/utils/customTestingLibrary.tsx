import '../app/global.css';
import React from 'react';
import { ToastContainer } from '@/components/Toast';
import { NextThemeProvider } from '@/providers/NextThemeProvider';
import { Provider as TooltipProvider } from '@radix-ui/react-tooltip';
import { render } from '@testing-library/react';

const customRender = (ui: React.ReactElement, { ...renderOptions } = {}) => render(
  <NextThemeProvider>
    <TooltipProvider>
      {ui}
      <ToastContainer />
    </TooltipProvider>
  </NextThemeProvider>,
  renderOptions
);

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';
export { customRender as render };
