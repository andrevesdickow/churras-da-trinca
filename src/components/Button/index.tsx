import * as React from 'react';
import { cva, type VariantProps } from 'class-variance-authority';
import { Loader } from 'lucide-react';
import { cn } from '@/lib/utils';
import { Slot } from '@radix-ui/react-slot';

const buttonVariants = cva(
  'inline-flex items-center justify-center gap-2 rounded-3xl text-sm font-medium uppercase ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-violet-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 dark:ring-offset-violet-950 dark:focus-visible:ring-violet-300',
  {
    variants: {
      variant: {
        default: 'bg-slate-950 text-white hover:bg-slate-950/90 dark:bg-slate-950 dark:text-white dark:hover:bg-slate-950/50',
        outline: 'border border-slate-950 bg-transparent text-slate-950 hover:bg-amber-100 dark:border-slate-950 dark:bg-transparent dark:hover:bg-amber-100',
        ghost: 'bg-transparent text-slate-950 hover:bg-amber-100 dark:bg-transparent dark:text-slate-300 dark:hover:bg-amber-400 dark:hover:text-slate-950',
        link: 'bg-transparent text-slate-950 hover:underline dark:bg-transparent dark:text-slate-950'
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-8 w-8'
      }
    },
    defaultVariants: {
      variant: 'default',
      size: 'default'
    }
  }
);

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> &
  VariantProps<typeof buttonVariants> & {
    asChild?: boolean;
    isLoading?: boolean;
  }

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ children, disabled, className, variant, size, asChild = false, isLoading = false, ...rest }, ref) => {
    const Comp = asChild ? Slot : 'button';
    const isDisabled = disabled || isLoading;

    return (
      <Comp
        disabled={isDisabled}
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...rest}
      >
        {children}
        {isLoading && <Loader className="h-4 w-4 animate-spin" />}
      </Comp>
    );
  }
);

Button.displayName = 'Button';

export { Button, buttonVariants };
