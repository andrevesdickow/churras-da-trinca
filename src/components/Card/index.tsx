import { cn } from '@/lib/utils';

/**
 * Component Composition Pattern
 */

type CardProps = React.ComponentProps<'div'>;

type CardHeaderProps = React.ComponentProps<'div'>;

type CardTitleProps = React.ComponentProps<'h2'>;

type CardSubtitleProps = React.ComponentProps<'h5'>;

type CardBodyProps = React.ComponentProps<'div'>;

const Card = ({ className, ...rest }: CardProps) => (
  <div
    className={cn('w-full h-full min-h-[160px] text-slate-800 dark:text-slate-50 bg-slate-50 dark:bg-slate-800 flex flex-col justify-between gap-6 p-6 rounded-lg shadow-xl', className)}
    {...rest}
  />
);

const CardHeader = ({ className, ...rest }: CardHeaderProps) => (
  <div
    className={cn('flex flex-col', className)}
    {...rest}
  />
);

const CardTitle = ({ className, ...rest }: CardTitleProps) => (
  <h2
    className={cn('text-2xl font-bold', className)}
    {...rest}
  />
);

const CardSubtitle = ({ className, ...rest }: CardSubtitleProps) => (
  <h5
    className={cn('text-lg font-semibold', className)}
    {...rest}
  />
);

const CardBody = ({ className, ...rest }: CardBodyProps) => (
  <div
    className={cn('flex justify-between', className)}
    {...rest}
  />
);

export { Card, CardHeader, CardTitle, CardSubtitle, CardBody };
