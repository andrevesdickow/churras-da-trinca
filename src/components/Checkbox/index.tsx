import * as React from 'react';
import { useController } from 'react-hook-form';
import toString from 'lodash/toString';
import { Check as CheckIcon } from 'lucide-react';
import { cn } from '@/lib/utils';
import * as CheckboxPrimitive from '@radix-ui/react-checkbox';
import { helperTextClassName } from '../TextField';

type CheckboxProps = React.ComponentPropsWithoutRef<typeof CheckboxPrimitive.Root> & {
  name: string;
  control: any;
  label?: React.ReactNode | string;
  helperText?: string;
}

const Checkbox = React.forwardRef<
  React.ElementRef<typeof CheckboxPrimitive.Root>,
  CheckboxProps
>(({
  className,
  name,
  control,
  label,
  helperText,
  checked = false,
  ...rest
}, ref) => {
  const { field, fieldState: { invalid, error } } = useController({
    name,
    control,
    defaultValue: checked
  });

  return (
    <div className="flex flex-col mb-2">
      <div className="flex gap-2 items-center">
        <CheckboxPrimitive.Root
          {...field}
          ref={ref}
          id={name}
          name={name}
          className={cn(
            'peer h-4 w-4 shrink-0 rounded-sm border border-amber-300 ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-violet-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-amber-300 data-[state=checked]:text-white dark:border-amber-300 dark:ring-offset-violet-950 dark:focus-visible:ring-violet-300 dark:data-[state=checked]:bg-amber-300 dark:data-[state=checked]:text-white',
            className
          )}
          value={toString(field.value)}
          defaultChecked={checked}
          onCheckedChange={(newChecked) => field.onChange(newChecked)}
          {...rest}
        >
          <CheckboxPrimitive.Indicator
            className={cn('flex items-center justify-center text-current')}
          >
            <CheckIcon className="h-4 w-4" />
          </CheckboxPrimitive.Indicator>
        </CheckboxPrimitive.Root>
        <label
          htmlFor={name}
          className="text-sm text-gray-500 dark:text-gray-300"
        >
          {label}
        </label>
      </div>
      {helperText || invalid && (
        <div className={helperTextClassName({ isInvalid: invalid })} role={invalid ? 'alert' : 'note'}>
          {invalid ? error?.message : helperText}
        </div>
      )}
    </div>
  );
});

Checkbox.displayName = CheckboxPrimitive.Root.displayName;

export { Checkbox };
