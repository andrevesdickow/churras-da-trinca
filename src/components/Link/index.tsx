import NextLink, { type LinkProps as NextLinkProps } from 'next/link';
import { cn } from '@/lib/utils';

type LinkProps = {
  children: React.ReactNode;
  className?: string;
} & NextLinkProps;

export const Link = ({ className, ...rest }: LinkProps) => (
  <NextLink
    className={cn('text-amber-400 hover:text-amber-500 transition-colors', className)}
    {...rest}
  />
);
