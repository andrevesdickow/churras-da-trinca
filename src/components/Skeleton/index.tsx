import { cn } from '@/lib/utils';

export const Skeleton = ({
  className,
  ...rest
}: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    role="status"
    className={cn('animate-pulse rounded-3xl bg-slate-100 dark:bg-slate-800', className)}
    {...rest}
  />
);
