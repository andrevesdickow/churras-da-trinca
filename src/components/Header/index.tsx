'use client';

import { useRouter } from 'next/navigation';
import { deleteCookie } from 'cookies-next';
import { LogOut as LogOutIcon } from 'lucide-react';
import { AuthKeys } from '@/enums/AuthKeys';
import { Button } from '../Button';
import { Tooltip, TooltipContent, TooltipTrigger } from '../Tooltip';

export const Header = () => {
  const router = useRouter();

  const handleLogout = () => {
    deleteCookie(AuthKeys.ACCESS_TOKEN);
    router.replace('/');
  };

  return (
    <header className="w-full h-[250px] fixed bg-gradient-to-b from-amber-400 via-70% via-amber-100 to-white dark:via-black/5 dark:to-black-900">
      <div className="bg-barbecue w-full h-full absolute" />
      <div className="absolute top-1 right-1">
        <Tooltip>
          <TooltipTrigger asChild>
            <Button
              type="button"
              onClick={handleLogout}
              variant="default"
              size="icon"
              aria-label="Sair"
            >
              <LogOutIcon className="w-4 h-4" />
            </Button>
          </TooltipTrigger>
          <TooltipContent side="left">
            Sair
          </TooltipContent>
        </Tooltip>
      </div>
      <h1 className="relative font-semibold text-center text-4xl mt-10 px-5 text-white dark:text-black">
        Churras da <span className="font-bold uppercase">Trinca</span>
      </h1>
    </header>
  );
};
