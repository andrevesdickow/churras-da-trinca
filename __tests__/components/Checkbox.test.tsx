import { useForm } from 'react-hook-form';
import { Checkbox } from '@/components/Checkbox';
import { render, screen, fireEvent, renderHook } from '@/utils/customTestingLibrary';

type CheckboxTestData = {
  checkboxTest: boolean;
};
const name = 'checkboxTest';

describe('<Checkbox />', () => {
  it('should render correctly Checkbox', () => {
    const { result: resultUseForm } = renderHook(() =>
      useForm<CheckboxTestData>()
    );

    const { control } = resultUseForm.current;

    const { container } = render(<Checkbox name={name} control={control} />);

    expect(screen.getByRole('checkbox')).not.toBeChecked();
    expect(container).toMatchSnapshot();
  });

  it('should render the checked correctly in Checkbox', () => {
    const { result: resultUseForm } = renderHook(() =>
      useForm<CheckboxTestData>()
    );

    const { control } = resultUseForm.current;

    render(<Checkbox name={name} control={control} />);

    fireEvent.click(screen.getByRole('checkbox'));

    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  it('should render the helper text correctly in Checkbox', async () => {
    const { result: resultUseForm } = renderHook(() =>
      useForm<CheckboxTestData>()
    );

    const { control } = resultUseForm.current;

    render(
      <Checkbox
        name={name}
        control={control}
        helperText="This field is required"
      />
    );

    expect(screen.getByText('This field is required')).toBeInTheDocument();
  });

  it('should render the label correctly in Checkbox', async () => {
    const { result: resultUseForm } = renderHook(() =>
      useForm<CheckboxTestData>()
    );

    const { control } = resultUseForm.current;

    render(<Checkbox name={name} label="Checkbox test" control={control} />);

    expect(screen.getByLabelText(/Checkbox test/)).toBeInTheDocument();
  });
});
