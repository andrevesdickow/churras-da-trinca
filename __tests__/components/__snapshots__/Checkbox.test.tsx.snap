// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> should render correctly Checkbox 1`] = `
<div>
  <script>
    !function(){try{var d=document.documentElement,c=d.classList;c.remove('light','dark');var e=localStorage.getItem('theme');if('system'===e||(!e&&true)){var t='(prefers-color-scheme: dark)',m=window.matchMedia(t);if(m.media!==t||m.matches){d.style.colorScheme = 'dark';c.add('dark')}else{d.style.colorScheme = 'light';c.add('light')}}else if(e){c.add(e|| '')}if(e==='light'||e==='dark')d.style.colorScheme=e}catch(e){}}()
  </script>
  <div
    class="flex flex-col mb-2"
  >
    <div
      class="flex gap-2 items-center"
    >
      <button
        aria-checked="false"
        class="peer h-4 w-4 shrink-0 rounded-sm border border-amber-300 ring-offset-white focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-violet-950 focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-amber-300 data-[state=checked]:text-white dark:border-amber-300 dark:ring-offset-violet-950 dark:focus-visible:ring-violet-300 dark:data-[state=checked]:bg-amber-300 dark:data-[state=checked]:text-white"
        data-state="unchecked"
        id="checkboxTest"
        role="checkbox"
        type="button"
        value="false"
      />
      <label
        class="text-sm text-gray-500 dark:text-gray-300 hover:cursor-pointer"
        for="checkboxTest"
      />
    </div>
  </div>
</div>
`;
